File: ocaml/fingertree.ml
type elem = Value of int64 | Node2 of elem * elem | Node3 of elem * elem * elem

type digit =
  | One of elem
  | Two of elem * elem
  | Three of elem * elem * elem
  | Four of elem * elem * elem * elem

type tree = Empty | Single of elem | Deep of digit * tree * digit

let node_to_digit = function
  | Node2 (a, b) -> Two (a, b)
  | Node3 (a, b, c) -> Three (a, b, c)
  | Value _ -> invalid_arg "node_to_digit"

let digit_to_tree = function
  | One a -> Single a
  | Two (a, b) -> Deep (One a, Empty, One b)
  | Three (a, b, c) -> Deep (Two (a, b), Empty, One c)
  | Four (a, b, c, d) -> Deep (Two (a, b), Empty, Two (c, d))

let rec snoc tree value =
  match tree with
  | Empty -> Single value
  | Single a -> Deep (One a, Empty, One value)
  | Deep (prefix, middle, Four (a, b, c, d)) ->
      Deep (prefix, snoc middle (Node3 (a, b, c)), Two (d, value))
  | Deep (prefix, middle, One a) -> Deep (prefix, middle, Two (a, value))
  | Deep (prefix, middle, Two (a, b)) -> Deep (prefix, middle, Three (a, b, value))
  | Deep (prefix, middle, Three (a, b, c)) ->
      Deep (prefix, middle, Four (a, b, c, value))

let rec view_left = function
  | Empty -> None
  | Single value -> Some (value, Empty)
  | Deep (One value, middle, suffix) -> (
      match view_left middle with
      | None -> Some (value, digit_to_tree suffix)
      | Some (node, middle') ->
          Some (value, Deep (node_to_digit node, middle', suffix)))
  | Deep (Two (a, b), middle, suffix) -> Some (a, Deep (One b, middle, suffix))
  | Deep (Three (a, b, c), middle, suffix) ->
      Some (a, Deep (Two (b, c), middle, suffix))
  | Deep (Four (a, b, c, d), middle, suffix) ->
      Some (a, Deep (Three (b, c, d), middle, suffix))

let modulus = 1_000_000_007L

let () =
  let rec build index tree =
    if index = 65_536 then tree
    else build (index + 1) (snoc tree (Value (Int64.of_int index)))
  in
  let rec churn round remaining tree checksum =
    if remaining = 0 then (tree, checksum)
    else
      match view_left tree with
      | Some (Value value, rest) ->
          let next = Int64.rem (Int64.add value (Int64.add round 1L)) modulus in
          churn (Int64.add round 1L) (remaining - 1) (snoc rest (Value next))
            (Int64.rem (Int64.add checksum value) modulus)
      | _ -> failwith "invalid finger tree"
  in
  let rec drain tree checksum =
    match view_left tree with
    | None -> checksum
    | Some (Value value, rest) ->
        drain rest (Int64.rem (Int64.add checksum value) modulus)
    | _ -> failwith "invalid finger tree"
  in
  let tree, checksum = churn 0L 1_000_000 (build 0 Empty) 0L in
  let result = drain tree checksum in
  if result <> 66_797_929L then (
    Printf.eprintf "FAIL: expected 66797929, got %Ld\n" result;
    exit 1)

File: rust/fingertree.rs
// Persistent finger tree. The constructors and snoc/view-left rotations
// mirror Data.FingerTree; Rc path-copying keeps every prior root valid.

use std::rc::Rc;

enum Elem {
    Value(i64),
    Node2(ElemRef, ElemRef),
    Node3(ElemRef, ElemRef, ElemRef),
}

type ElemRef = Rc<Elem>;

enum Digit {
    One(ElemRef),
    Two(ElemRef, ElemRef),
    Three(ElemRef, ElemRef, ElemRef),
    Four(ElemRef, ElemRef, ElemRef, ElemRef),
}

enum Tree {
    Empty,
    Single(ElemRef),
    Deep(Digit, TreeRef, Digit),
}

type TreeRef = Rc<Tree>;

fn empty() -> TreeRef { Rc::new(Tree::Empty) }
fn value(value: i64) -> ElemRef { Rc::new(Elem::Value(value)) }
fn node3(a: ElemRef, b: ElemRef, c: ElemRef) -> ElemRef { Rc::new(Elem::Node3(a, b, c)) }

fn node_to_digit(node: &ElemRef) -> Digit {
    match node.as_ref() {
        Elem::Node2(a, b) => Digit::Two(a.clone(), b.clone()),
        Elem::Node3(a, b, c) => Digit::Three(a.clone(), b.clone(), c.clone()),
        Elem::Value(_) => panic!("invalid middle-tree element"),
    }
}

fn digit_to_tree(digit: &Digit) -> TreeRef {
    match digit {
        Digit::One(a) => Rc::new(Tree::Single(a.clone())),
        Digit::Two(a, b) => Rc::new(Tree::Deep(
            Digit::One(a.clone()), empty(), Digit::One(b.clone()),
        )),
        Digit::Three(a, b, c) => Rc::new(Tree::Deep(
            Digit::Two(a.clone(), b.clone()), empty(), Digit::One(c.clone()),
        )),
        Digit::Four(a, b, c, d) => Rc::new(Tree::Deep(
            Digit::Two(a.clone(), b.clone()), empty(), Digit::Two(c.clone(), d.clone()),
        )),
    }
}

fn snoc(tree: &TreeRef, x: ElemRef) -> TreeRef {
    match tree.as_ref() {
        Tree::Empty => Rc::new(Tree::Single(x)),
        Tree::Single(a) => Rc::new(Tree::Deep(
            Digit::One(a.clone()), empty(), Digit::One(x),
        )),
        Tree::Deep(prefix, middle, Digit::Four(a, b, c, d)) => Rc::new(Tree::Deep(
            clone_digit(prefix),
            snoc(middle, node3(a.clone(), b.clone(), c.clone())),
            Digit::Two(d.clone(), x),
        )),
        Tree::Deep(prefix, middle, Digit::One(a)) => Rc::new(Tree::Deep(
            clone_digit(prefix), middle.clone(), Digit::Two(a.clone(), x),
        )),
        Tree::Deep(prefix, middle, Digit::Two(a, b)) => Rc::new(Tree::Deep(
            clone_digit(prefix), middle.clone(), Digit::Three(a.clone(), b.clone(), x),
        )),
        Tree::Deep(prefix, middle, Digit::Three(a, b, c)) => Rc::new(Tree::Deep(
            clone_digit(prefix), middle.clone(), Digit::Four(a.clone(), b.clone(), c.clone(), x),
        )),
    }
}

fn clone_digit(digit: &Digit) -> Digit {
    match digit {
        Digit::One(a) => Digit::One(a.clone()),
        Digit::Two(a, b) => Digit::Two(a.clone(), b.clone()),
        Digit::Three(a, b, c) => Digit::Three(a.clone(), b.clone(), c.clone()),
        Digit::Four(a, b, c, d) => Digit::Four(a.clone(), b.clone(), c.clone(), d.clone()),
    }
}

fn view_left(tree: &TreeRef) -> Option<(ElemRef, TreeRef)> {
    match tree.as_ref() {
        Tree::Empty => None,
        Tree::Single(x) => Some((x.clone(), empty())),
        Tree::Deep(Digit::One(x), middle, suffix) => {
            let rest = match view_left(middle) {
                None => digit_to_tree(suffix),
                Some((node, middle_rest)) => Rc::new(Tree::Deep(
                    node_to_digit(&node), middle_rest, clone_digit(suffix),
                )),
            };
            Some((x.clone(), rest))
        }
        Tree::Deep(Digit::Two(a, b), middle, suffix) => Some((
            a.clone(),
            Rc::new(Tree::Deep(Digit::One(b.clone()), middle.clone(), clone_digit(suffix))),
        )),
        Tree::Deep(Digit::Three(a, b, c), middle, suffix) => Some((
            a.clone(),
            Rc::new(Tree::Deep(
                Digit::Two(b.clone(), c.clone()), middle.clone(), clone_digit(suffix),
            )),
        )),
        Tree::Deep(Digit::Four(a, b, c, d), middle, suffix) => Some((
            a.clone(),
            Rc::new(Tree::Deep(
                Digit::Three(b.clone(), c.clone(), d.clone()), middle.clone(), clone_digit(suffix),
            )),
        )),
    }
}

fn main() {
    const P: i64 = 1_000_000_007;
    let mut tree = empty();
    for x in 0..65_536 { tree = snoc(&tree, value(x)); }
    let mut checksum = 0;
    for round in 0..1_000_000i64 {
        let (element, rest) = view_left(&tree).expect("non-empty tree");
        let x = match element.as_ref() { Elem::Value(x) => *x, _ => unreachable!() };
        checksum = (checksum + x) % P;
        tree = snoc(&rest, value((x + round + 1) % P));
    }
    while let Some((element, rest)) = view_left(&tree) {
        let x = match element.as_ref() { Elem::Value(x) => *x, _ => unreachable!() };
        checksum = (checksum + x) % P;
        tree = rest;
    }
    if checksum != 66_797_929 {
        eprintln!("FAIL: expected 66797929, got {}", checksum);
        std::process::exit(1);
    }
}
